triage/summarize/map_abstractions.go

```diff
@@ -0,0 +1,209 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+/*
+This file contains abstractions of certain complex map types that are used often throughout
+summarize.go. They also have some often-used methods such as getting just the keys, or getting the
+key-value pairs as a map.
+*/
+
+package summarize
+
+import (
+	"sort"
+)
+
+// map[string][]failure type alias
+
+// failuresGroup maps strings to failure slices.
+type failuresGroup map[string][]failure
+
+// failuresGroupPair is a representation of a failuresGroup key-value mapping as a two-element
+// struct.
+type failuresGroupPair struct {
+	Key      string    `json:"key"`
+	Failures []failure `json:"failures"`
+}
+
+// keys provides the failuresGroup's keys as a string slice.
+func (fg *failuresGroup) keys() []string {
+	result := make([]string, len(*fg))
+
+	iter := 0
+	for key := range *fg {
```

Comment:
if you know that len(*fg) is the len of the array, I think you can just
```
 result := make([]string, 0, len(*fg))
for key := range *fg {
   result = append(result, key)
}
```

Reply:
Addressed in #18882